2018/001-external-authentication.md

```diff
@@ -0,0 +1,125 @@
+# External Authentication Integration
+
+| Field  | Value      |
+|:-------|:-----------|
+| Status | Draft      |
+| Date   | 2018-05-03 |
+
+## Introduction
+
+### Problem description
+
+Many enterprises will have a pre-existing authentication service such as LDAP,
+Active Directory, SAML 2.0, etc and wish to use it to handle authentication of
+users for Kubernetes.
+
+### Proposed change
+
+* Allow operators to provide connection details for external authentication
+  systems via Velum.
+* Configure Dex with these connection details, enabling Kubernetes to
+  authenticate, via Dex, against these external services.
+* Support external LDAP, SAML 2.0 and OpenID Connect authentication
+  services.
+
+## Detailed RFC
+
+Kubic/CaaSP currently uses Dex as an authentication intermediary for
+Kubernetes, where Dex is backed by a Kubic managed LDAP server.
+
+Dex supports many authentication "connectors" - LDAP, SAML 2.0, OpenID Connect,
+as well as several less common connectors such as GitHub, LinkedIn, etc. Dex
+additionally supports configuring multiple connectors at once, allowing users
+to choose which service they need to authenticate against.
+
+This RFC proposes allowing Kubic operators to configure arbitary additional
+connectors within Kubic, pointing towards their own authentication backend.
+Initially, this will be limited to Dex's LDAP, SAML 2.0 and OpenID Connect
+connectors.
+
+This RFC does not propose removing that managed LDAP server, though that may
+be something we wish to pursue in the future.
+
+### Proposed change
+
+#### Velum Workflow Changes
+
+We propose to leave the initial bootstrap workflow unchanged, and instead
+allow additional authentication connectors to be configured post-bootstrap, in
+a similar manner to how we allow containter registries to be configured
+post-bootstap.
+
+Operators should be able to login to Velum, navigate to the Settings tab, and
+view, add or remove authentication connectors. Each authentication connector
+type (LDAP, SAML 2.0, OpenID Connect) will have a different view for the view
+and add pages, as Dex requires a different set of configuration inputs for
+each.
+
+It should be possible for operators to configure multiple connectors
+simultaneously, including multiple connectors of the same type (e.g. 2x LDAP).
+
+Once connectors have been configured, Velum will show an "Apply Changes" button,
```

Comment:
Please also include a "test" button that has to validate successfully before the "Apply Changes" button is activated

Reply:
In order to actually do this, it means we'd need to implement clients for each of the protocols.. This is likely to be problematic as - it's a chunk of extra work, and the validations wouldn't actually be using the same code as a the real time, nor would the connections be coming from the same place. Nothing insurmountable, but, certainly a much larger chunk of work.

I think we can initially handle this by requesting that operators create a second auth backend, "Acme Corp LDAP - Test" - which can be provisioned and then tested as functional - prior to copying the settings over. It's not pretty, but might allow enough flexibility to validate prior to migrating all users over to the new settings?

Down the line, we could possibly add the ability to deploy a second copy of Dex, along with a test application. This way, operators could visit the temporary Dex URL (likely just exposed as a NodePort on the cluster), login, and be sent the the test application. The test application would provide a dump of the relevant authentication data - user, groups, etc.